salesagility/SuiteCRM
# Fix #7854 Adds authorization code grant type flow to Api V8 OAuth2

As discussed in #7854 this adds the OAuth2 authorization code grant type flow to the Api V8. The code leverages thephpleague's OAuth2 Server and follows the [implementation guidelines](https://oauth2.thephpleague.com/authorization-server/auth-code-grant/).

The PR covers the following main changes:
- It adds a new module to store OAuth2 authorization codes. The module key is ```OAuth2AuthCodes```
- It implements the authorization code grant flow in the Api V8 middleware and stores the authorization codes in the newly created OAuth2AuthCodes module

The OAuth2 authorization code flow includes the following steps:
1. A client redirects the user to the SuiteCRM instance (path-to-instance/Api/authorize) with a client_id and redirect_uri parameter.
2. SuiteCRM validates that the client_id and the redirect_uri correspond to a valid OAuth2 Client and redirects the user to the login page (if not logged-in already). Once logged-in the user is requested to authorize the client. This takes place on a custom view (view.authorize.php) in the new module. The user can:
a. decline the request, which redirects the user back to the client (the redirect contains an approapriate hint)
b. authorize the client, which creates a new authorization code. The code is saved as a new record in the OAuth2AuthCode module. The user is redirected to the client. The redirect url contains the authorization code.
3. The client than sends a direct request containing the authorization code to SuiteCRMs ```Api/access_token``` access point. The OAuth2 Server then again validates the client and additionally the client secret and the authorization code. If both are valid and match a new access_token on behalf of the authorizing user is returned. This access_token now grants the client access to the V8 Api. The client thereby acts as the authorizing user. 

## Motivation and Context
Access to the V8 Api is currently granted by either a client credentials grant or a password grant. The client credentials grant has the downside, that it allows to access the Api on behalf of a dedicated user  only (the user is defined in the admin area). The password grant has the downside that the client system needs to handle the user's SuiteCRM password to obtain access to the Api, which is a potential security risk. The authoriazion code grant solves these issues.

## Additional feature and implementation description ##
4. In line with similar implementations in other systems, we offer the user a third option (besides authorizing and declining the request): The user can choose to save the authorization decision for future requests. We thenflag the OAuth2AuthCode record. The flag first prevents that the record is deleted once the authorization code is expired (the code though does expire). Second, if a new authorization code is requested by that particular client AND if the corresponding user is logged-in into SuiteCRM a new authorization code is returned without showing the decision page. The authorize view looks as follows (my color scheme is blue...): 
![grafik](https://user-images.githubusercontent.com/23050940/69373554-5a74df00-0ca4-11ea-813e-7355ef64bc7c.png)

5. The OAuth2AuthCodes module appears in the module list, because users can access the list view. For normal users, the list view lists the authorization codes for that particular user. For admins, the list view lists all authorization codes. The list view is accessible for normal users, because they can revoke their authorization codes there. This is in particular relevant for the authoriation codes that have been saved for automatic approval of future requests (see point 4). 
![grafik](https://user-images.githubusercontent.com/23050940/69373800-ed157e00-0ca4-11ea-8be8-b0bd0ef2f9b5.png)

6. If a client redirects a user to the SuiteCRM instance (to obtain an auth code) and if that user is not logged-in already, the user will be redirected to the login page (the user logs-in in order to obtain the authorization code for the client). In this case, the user is [logged-out automatically](https://github.com/HVStechnik/SuiteCRM/blob/d9ea92abc86436c46e9d1dc75f1fe42f0f4feb37/modules/OAuth2AuthCodes/views/view.authorize.php#L131-L133) once the authorization request has been approved or declined. This is because we do not want to keep a logged-in SuiteCRM session, the user might not be aware of, because once the user is redirected to the client, there is no open SuiteCRM window any more.

## How To Test This
- The OAuth2AuthCode module should be visible, if activated in the menu.
- Run the Api V8 activation robo task
- Create a new authoriation code client in the admin OAuth2 Clients section
- Ensure that the [htaccess RewriteRule](https://github.com/HVStechnik/SuiteCRM/blob/d9ea92abc86436c46e9d1dc75f1fe42f0f4feb37/install/install_utils.php#L1018) is in place.
- Access: ```path-to-instance/Api/authorize?response_type=code&client_id=[client identifier]&redirect_uri=[redirect_url]```
- Once an authoriation code is returned, you can use this auth code with the client id and the client secret to obtain an access token. This can be tested in Postman.

Please ask, if there are any questions! I'm happy to furhter explain and improve the code in case of any recommendations.

## Types of changes


### Final checklist
